### ISSUE TITLE
Serializing JIT compilation

### ISSUE THREAD

**Issue by soufianekhiat** [halide/Halide#6116]
Is they is a way to serialize a jit compilation?

Use case:
```cpp
// Build the Func f;
Halide::Target target = Halide::get_host_target();
uint64_t hash = HashFunc( f, target ); // TODO
Pipeline* pPipeline = nullptr;
if ( ExistOnDisk( hash ) ) // TODO
{
    pPipeline = GetPipelineFromDisk( hash ); // TODO
}
else
{
    pPipeline = Pipeline( f );
    pPipeline->auto_schedule( GetCurrentAutoSchedule(), target, Halide::MachineParams( GetCoreCount(), IsGPU() ? GetGPUMemory() : GetCPUMemory(), GetBalance( IsGPU() ) ) );
    pPipeline->compile_jit( target );
    StoreOnDisk( hash, f ); // TODO
}
```

I try to look on Module but all data are read-only.

**Comment by abadams** [comment]
I wish llvm could emit shared objects directly. The only way to do this now I think requires compiling to an object, then shelling out to a linker to turn that into a .so, and then dlopening the .so.

**Comment by alexreinking** [comment]
> I wish llvm could emit shared objects directly. The only way to do this now I think requires compiling to an object, then shelling out to a linker to turn that into a .so, and then dlopening the .so.

Is there not a way to do this by linking to lld? We do something like this for lldWasm, don't we?

**Comment by abadams** [comment]
I was under the impression lld was not available as a library, just as a command-line tool.

**Comment by steven-johnson** [comment]
> I was under the impression lld was not available as a library, just as a command-line tool.

Nope, it's totally available as a library. We use lldWasm in this way.

**Comment by abadams** [comment]
Might be cool to support shared library as an output type then, as it would enable the sort of thing requested in this issue.

**Comment by mikewoodworth** [comment]
One option might be to emit bytecode of the module post optimization... though I have no idea what percentage of the overall JIT time that would save. I know Apple did something similar for OpenCL kernels in the past.

**Comment by soufianekhiat** [comment]
After a try I maybe missed something.
I produced an obj with: `compile_to_object`.
And linked it with:
lld-link.exe /defaultlib:libcmt.lib /wholearchive /machine:x64 /subsystem:windows /dll dumpObj /out:dumpObj.dll

Note: Currently my target is Windows only.

Checking the format of the obj generated:
```
File: dumpObj
Format: COFF-x86-64
Arch: x86_64
AddressSize: 64bit
```

By checking the content of the generated DLL I don't see my function which is present on the object by checking with:

**Comment by alexreinking** [comment]
> By checking the content of the generated DLL I don't see my function which is present on the object by checking with:

@soufianekhiat - it looks like your comment got cut off?

**Comment by soufianekhiat** [comment]
@alexreinking 
my bad fixed.

**Comment by soufianekhiat** [comment]
After trying to dump differently with `/wholearchive:dumpObj.lib`:
```
lld-link.exe /defaultlib:libcmt.lib /implib:dumpObj.lib /wholearchive /wholearchive:dumpObj.lib /machine:x64 /subsystem:windows /dll dumpObj.obj /out:dumpObj.dll
lld-link: error: dumpObj.lib(dumpObj.dll): .idata$4 should not refer to special section 0
```

**Comment by soufianekhiat** [comment]
Solution add export by hand with option:
`/wholearchive /include:coscos /export:coscos`
Relevant ref:
https://stackoverflow.com/a/58913687
Ideally we would like either output directly a dll or a def file:
https://stackoverflow.com/a/62360059
